restaking/src/symbiotic/mod.rs

```diff
@@ -0,0 +1,250 @@
+            network_middleware_address,
+            state_retriever_address,
+        }
+    }
+
+    pub async fn get_operators(
+        &self,
+        timestamp: Option<u64>,
+    ) -> Result<HashSet<Address>, alloy::contract::Error> {
+        let network_middleware =
+            MiddlewareReader::new(self.network_middleware_address, self.provider.clone());
+
+        let addresses = match timestamp {
+            None => network_middleware.activeOperators().call().await?._0,
+            Some(timestamp) => {
+                let timestamp = Uint::<48, 1>::from(timestamp);
+                network_middleware
+                    .activeOperatorsAt(timestamp)
+                    .call()
+                    .await?
+                    ._0
+            }
+        };
+
+        let mut set = HashSet::new();
+        for address in addresses {
+            set.insert(address);
+        }
+        Ok(set)
+    }
+
+    pub async fn get_validator_set(
+        &self,
+        timestamp: Option<u64>,
+    ) -> Result<Vec<Validator>, alloy::contract::Error> {
+        let state_retriever =
+            StateRetriever::new(self.state_retriever_address, self.provider.clone());
+
+        let validators_data = match timestamp {
+            None => {
+                state_retriever
+                    .getValidatorSet(self.network_middleware_address)
+                    .call()
+                    .await?
+                    .validatorSet
+            }
+            Some(timestamp) => {
+                let timestamp = Uint::<48, 1>::from(timestamp);
+                state_retriever
+                    .getValidatorSetAt(self.network_middleware_address, timestamp)
+                    .call()
+                    .await?
+                    .validatorSet
+            }
+        };
+
+        let mut validators = Vec::with_capacity(validators_data.len());
+        for data in validators_data {
+            validators.push(Validator {
+                operator: data.operator,
+                key: data.key,
+                power: data.power,
+            });
+        }
+        Ok(validators)
+    }
+}
+
+pub struct Validator {
+    pub operator: Address,
+    pub key: FixedBytes<32>,
+    pub power: U256,
+}
+
+#[cfg(test)]
+mod tests {
+    use alloy::{primitives::FixedBytes, primitives::U256, providers::ProviderBuilder, sol};
+    use alloy_node_bindings::Anvil;
+
+    use super::*;
+    use rand::Rng;
+
+    // Codegen from compiled MockedMiddlewareReader contract.
+    sol!(
+        #[allow(missing_docs)]
+        #[sol(rpc)]
+        MockedNetworkMiddleware,
+        "src/symbiotic/artifacts/MockedMiddlewareReader.sol/MockedMiddlewareReader.json"
+    );
+
+    // Codegen from compiled MockedStateRetriever contract.
+    sol!(
+        #[allow(missing_docs)]
+        #[sol(rpc)]
+        MockedStateRetriever,
+        "src/symbiotic/artifacts/MockedStateRetriever.sol/MockedStateRetriever.json"
+    );
+
+    #[tokio::test]
```

Comment:
Before we can merge this, we'd want to migrate this to run in runtime::deterministic rather than tokio.

Is there an assumption that `tokio` is used inside of alloy or did you just use it for `.await`?

Reply:
It has been put because of the `await`, It required to have a test function being defined with `async` keyword but it is prohibited without puting attibute
`#[tokio::test]`.